### PR TITLE
Vectorise replacement function

### PR DESCRIPTION
**Pull request by hadley**
This breaks 7 revdeps but I think it's worth it because the performance improvement is immense.

### CODE REVIEW

**Comment by hadley** [comment]
It also breaks rmarkdown 😬 

```
Error in `if (grepl("^[.][.]", in_file)) ...`:
! the condition has length > 1
---
Backtrace:
     ▆
  1. └─pkgdown::build_article("from-base", new_process = FALSE)
  2.   └─pkgdown:::render_rmarkdown(...) at pkgdown/R/build-articles.R:281:2
  3.     ├─rlang::inject(rmarkdown::render(!!!args)) at pkgdown/R/rmarkdown.R:52:4
  4.     └─rmarkdown::render(...) at rlang/R/eval.R:195:2
  5.       └─output_format$post_processor(...)
  6.         └─rmarkdown (local) base(metadata, input_file, output_file, ...)
  7.           └─rmarkdown:::process_images(output_str, image_relative)
  8.             └─rmarkdown:::process_html_res(...)
  9.               └─stringr::str_replace_all(html, reg, process_img_src)
 10.                 └─stringr:::str_transform_all(string, pattern, replacement) at stringr/R/replace.r:91:4
 11.                   └─rmarkdown (local) replacement(old_flat) at stringr/R/replace.r:201:4
 12.                     └─rmarkdown (local) processor(img_src, src)
```

**Comment by hadley** [comment]
rmarkdown patch in https://github.com/rstudio/rmarkdown/pull/2416. Breaking rmarkdown is a big deal, so I don't think we should merge this until after rmarkdown is released, so it probably won't make it for this release.

**Comment by DavisVaughan** [comment]
I guess pkgdown is failing because rmarkdown is broken? Very meta

File: R/replace.R
Comment (DavisVaughan) [inline R/replace.R:226]:
```suggestion
  start <- cumsum(c(1L, ls[-length(ls)]))
  end <- start + ls - 1L
```

File: R/replace.R
Comment (DavisVaughan) [inline R/replace.R:195]: Depending on how much you value readability over performance, you don't actually need to pass `idx` here since you are unchopping in the order the input was provided

File: R/replace.R
Comment (DavisVaughan) [inline R/replace.R:199]: 💀

File: R/view.R
Comment (DavisVaughan) [inline R/view.R:115]: I guess this is where you happened to notice the cli issue 😆

File: tests/testthat/_snaps/replace.md
Comment (DavisVaughan) [inline tests/testthat/_snaps/replace.md:43]: My brain kind of broke on this because `replacement` is really the arg name, but i see what you are trying to do by adding `()` since it is used as a function call. I don't mind either way, just wanted to say it took me a second to realize it wasn't a typo

File: R/replace.R
Comment (DavisVaughan) [reply R/replace.R:199]: Alternate idea is to actually apply `replacement()` and then check if `length(new_flat) == 1 && length(old_flat) == 0`. In that special case you could issue a warning (possibly even mentioning `paste0()`) and "fix" it for them.

**Review by DavisVaughan — APPROVED** [review]
I guess the broken packages are like rmarkdown? They didn't expect multiple strings at once?

File: R/replace.R
Comment (hadley) [reply R/replace.R:199]: Yeah, we could do that, I just don't think it's worth the effort since we do know what the correct answer is in the 0 element case.

File: tests/testthat/_snaps/replace.md
Comment (hadley) [reply tests/testthat/_snaps/replace.md:43]: I'll tweak this to make it clearer.

**Comment by hadley** [comment]
Yeah, the failures seem to be cases that expected a single string. So there's always a trivial fix, which is to vectorise with `vapply()` or similar. When I do the revdep fixes, I'll try and vectorise properly so they can benefit from the performance boost.

**Comment by hadley** [comment]
Fixed in rmarkdown 2.17, but I think this will have to wait until the next release since otherwise installing stringr will break older (i.e. most) rmarkdown installed.

File: R/replace.R
Comment (hadley) [reply R/replace.R:195]: In dev vctrs:

```R
xs <- list("a", c("b", "c"), c("d", "e", "f"))

x <- list_unchop(xs)
x <- toupper(x)
x
#> [1] "A" "B" "C" "D" "E" "F"

sizes <- list_sizes(xs)
vec_partition(x, sizes)
```

**Comment by hadley** [comment]
I think enough time has elapsed that we can now merge this, fixing the 7 CRAN failures with PRs.

File: R/replace.R
Comment (hadley) [inline R/replace.R:203]: @DavisVaughan do you think it's worth wrapping this to provide a more informative error? (something like `replacement` function must be vectorised?)

Comment (DavisVaughan) [reply R/replace.R:203]: Are you talking about removing the `length(old_flat) == 0` special case handling?

That was what I mentioned back here, I think 
https://github.com/tidyverse/stringr/pull/462#discussion_r987127244

We went through a similar thing with `dplyr::rename_with()` where we made it so that `.fn` has to return a vector of length equal to the input length (i.e. typical size stability, like here).

For dplyr, we did get a few questions about this, so we added a `paste0(recycle0 = TRUE)` example:
https://github.com/tidyverse/dplyr/pull/6690

I think you could do a few things to mitigate issues:
- Add an example of `paste0`, like dplyr
- You already have a error below when `length(new_flat) != length(old_flat)`, but you could specially call out that `replacement` needs to be _vectorized_ there as well.
- If you're feeling particularly friendly, you could check to see if `replacement` was identical to `paste` or `paste0` in the check below after `length(new_flat) != length(old_flat)` fails. If so, you could throw a special error message just for that very common case, and probably advise to use `recycle0`.

Comment (hadley) [reply R/replace.R:203]: If you look at the revdeps, a lot of the errors are folks passing vectors to functions that expect scalars. So I was thinking it might make sense to wrap the call to `replacement()` to provide some hint in the error message.

Comment (DavisVaughan) [reply R/replace.R:203]: Ah like

```r
    Error in if (x > 0) "+" else "-" : the condition has length > 1
    Calls: <Anonymous> ... <Anonymous> -> str_transform_all -> replacement -> <Anonymous>
```

Ok now I understand what you mean.

I think it would be quite hard to make that error precise, i.e. you can't tell when they've provided a function that only expects a scalar very easily.

So the alternative is to _always_ wrap any error from `replacement()` with some general advice that talks needing a vectorized function. I think that could be useful if you frame it like:

```
Error in str_fn():
Failed to apply `replacement()`. Check the following in addition to the error thrown by `replacement()`:
- `replacement()` must be a vectorized function capable of handling input of any length.
- `replacement()` must return a character vector with output the same length as the input.
Caused by error in replacement():
<parent error here>
```

Comment (hadley) [reply R/replace.R:203]: Yeah, exactly!

**Pull request merged**
